pallets/offchain-demo/src/lib.rs

```diff
@@ -171,7 +204,85 @@ impl<T: Trait> Module<T> {
 		}
 	}
 
-	fn fetch_n_parse(_block_number: T::BlockNumber) -> Result<(), Error<T>> {
+	/// Check if we have fetched github info before. If yes, we use the cached version that is
+	///   stored in off-chain worker storage `storage`. If no, we fetch the remote info and then
+	///   write the info into the storage for future retrieval.
+	fn fetch_if_needed() -> Result<(), Error<T>> {
+
+		// Start off by creating a reference to Local Storage value.
+		// Since the local storage is common for all offchain workers, it's a good practice
+		// to prepend our entry with the pallet name.
+		let s_info = StorageValueRef::persistent(b"offchain-demo::gh-info");
+		let s_lock = StorageValueRef::persistent(b"offchain-demo::lock");
+
+		// The local storage is persisted and shared between runs of the offchain workers,
+		// and offchain workers may run concurrently. We can use the `mutate` function, to
+		// write a storage entry in an atomic fashion.
+		//
+		// It has a similar API as `StorageValue` that offer `get`, `set`, `mutate`.
+		// If we are using a get-check-set access pattern, we likely want to use `mutate` to access
+		// the storage in one go.
+		//
+		// Ref: https://substrate.dev/rustdocs/v2.0.0-alpha.6/sp_runtime/offchain/storage/struct.StorageValueRef.html
+		if let Some(Some(gh_info)) = s_info.get::<GithubInfo>() {
+			// gh-info has already been fetched. Return early.
+			debug::info!("cached gh-info: {:?}", gh_info);
+			return Ok(());
+		}
+
+		// Lock acquisition
+		let res: Result<Result<bool, bool>, Error<T>> = s_lock.mutate(|s: Option<Option<bool>>| {
+			match s {
+				// if the lock has never been set or is free (false), return true to execute `fetch_n_parse`
+				None | Some(Some(false)) => Ok(true),
+				// otherwise, someone already hold the lock (true), we want to skip `fetch_n_parse`.
+				_ => Err(<Error<T>>::AlreadyFetched),
+			}
+		});
+
+		// `res` cases:
+		//   Err() - someone hold the lock already, so we want to skip
+		//   Ok(Err(true)) - Another ocw is writing to the storage while we set it, so we also skip `fetch_n_parse` in this case.
+		//   Ok(Ok(true)) - we get the lock, so we run `fetch_n_parse`
+		if let Ok(Ok(true)) = res {
+			// Release the lock here in case `fetch_n_parse` fail and return early.
+			s_lock.set(&false);
+			let gh_info = Self::fetch_n_parse()?;
+			s_info.set(&gh_info);
+			debug::info!("fetched gh-info: {:?}", gh_info);
+		}
+
+		Ok(())
+
```

Comment:
@tomusdrw This is my attempt on writing a lock-acquiring mechanism to write to storage. Could you review and see if this implementation is good? Thanks.